server/terminal/terminal.go

```diff
@@ -0,0 +1,236 @@
+package terminal
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/argoproj/gitops-engine/pkg/utils/kube"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	applisters "github.com/argoproj/argo-cd/v2/pkg/client/listers/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/server/application"
+	servercache "github.com/argoproj/argo-cd/v2/server/cache"
+	"github.com/argoproj/argo-cd/v2/server/rbacpolicy"
+	"github.com/argoproj/argo-cd/v2/util/argo"
+	"github.com/argoproj/argo-cd/v2/util/db"
+	"github.com/argoproj/argo-cd/v2/util/rbac"
+)
+
+type terminalHandler struct {
+	appLister         applisters.ApplicationNamespaceLister
+	db                db.ArgoDB
+	enf               *rbac.Enforcer
+	cache             *servercache.Cache
+	appResourceTreeFn func(ctx context.Context, app *appv1.Application) (*appv1.ApplicationTree, error)
+}
+
+// NewHandler returns a new terminal handler.
+func NewHandler(appLister applisters.ApplicationNamespaceLister, db db.ArgoDB, enf *rbac.Enforcer, cache *servercache.Cache,
+	appResourceTree application.AppResourceTreeFn) *terminalHandler {
+	return &terminalHandler{
+		appLister:         appLister,
+		db:                db,
+		enf:               enf,
+		cache:             cache,
+		appResourceTreeFn: appResourceTree,
+	}
+}
+
+func (s *terminalHandler) getApplicationClusterRawConfig(ctx context.Context, a *appv1.Application) (*rest.Config, error) {
+	if err := argo.ValidateDestination(ctx, &a.Spec.Destination, s.db); err != nil {
+		return nil, err
+	}
+	clst, err := s.db.GetCluster(ctx, a.Spec.Destination.Server)
+	if err != nil {
+		return nil, err
+	}
+	return clst.RawRestConfig(), nil
+}
+
+func (s *terminalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	podName := q.Get("pod")
+	container := q.Get("container")
+	app := q.Get("appName")
+	namespace := q.Get("namespace")
+	shell := q.Get("shell")
+
+	if podName == "" || container == "" || app == "" || namespace == "" {
+		http.Error(w, "Missing required parameters", http.StatusBadRequest)
+		return
+	}
+
+	ctx := r.Context()
+	a, err := s.appLister.Get(app)
+	if err != nil {
+		http.Error(w, "Cannot get app", http.StatusBadRequest)
+		return
+	}
+
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplications, rbacpolicy.ActionGet, appRBACName(*a)); err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
```

Comment:
```suggestion
	a, err := s.appLister.Get(app)
	if err == nil {
		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplications, rbacpolicy.ActionGet, appRBACName(*a)); err != nil {
			http.Error(w, err.Error(), http.StatusUnauthorized)
			return
		}
	} else {
		http.Error(w, "Cannot get app", http.StatusBadRequest)
		return
	}
```

I think we should avoid leaking even the existence/non-existence of an app to a user who does not have permission to `get` that app.

Reply:
Thanks. I think I got your point and it makes sense to me. But I didn't get the change you made. Isn't it the same as the original code?